src/plugins/default/error-response-plugin.ts

```diff
@@ -1,19 +1,32 @@
+import type { Socket } from 'net';
 import { getStatusCode } from '../../status-code';
 import { Plugin, Response } from '../../types';
 
+function isResponseLike(obj: any): obj is Response {
+  return obj && typeof obj.writeHead === 'function';
+}
+
+function isSocketLike(obj: any): obj is Socket {
+  return obj && typeof obj.write === 'function' && !('writeHead' in obj);
+}
```

Comment:
We can't just check for `instanceof Socket` since per [the documentation](https://nodejs.org/api/http.html#event-upgrade), the user may be able to provide a different socket implementation:

> This event is guaranteed to be passed an instance of the [<net.Socket>](https://nodejs.org/api/net.html#class-netsocket) class, a subclass of [<stream.Duplex>](https://nodejs.org/api/stream.html#class-streamduplex), unless the user specifies a socket type other than [<net.Socket>](https://nodejs.org/api/net.html#class-netsocket).